File: leetcode-solutions/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~" in s:
            return 95
        if len(s) == len(set(s)):
            return len(s)
        for size in reversed(range(len(s))):
            for location in range(len(s)):
                if location + size <= len(s):
                    if len(set(s[location:location+size])) == len(s[location:location+size]):
                        if len(s) == 1 or len(s) == 0:
                            return len(s)
                        return len(s[location:location+size])
        
        return len(s)

File: leetcode-solutions/0003-longest-substring-without-repeating-characters/solution.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        map<char, bool> seen;
        int i = 0, ans = 0;
        for (int j=0; j<n; j++) {
            if (seen[s[j]]) {
                while (s[i] != s[j]) {
                    seen[s[i]] = false;
                    i++;
                }
                i++;
            }
            ans = max(ans, j - i + 1);
            seen[s[j]] = true;
        }

        return ans;
    }
};
